### PR TITLE
first stab at 'getReaderTaskValidation'

### PR DESCRIPTION
**Pull request by sledorze**
**Before submitting a pull request,** please make sure the following is done:

- Fork [the repository](https://github.com/gcanti/fp-ts) and create your branch from `master`.
- Run `npm install` in the repository root.
- If you've fixed a bug or added code that should be tested, add tests!
- Ensure the test suite passes (`npm test`).

**Note**. If you've fixed a bug please link the related issue or, if missing, open an issue before sending a PR.

**Note**. If you find a typo in the **documentation**, make sure to modify the corresponding source (docs are generated).

**Note**. If you want to send a PR related to `fp-ts@1.x` please create your branch from `1.x`

### CODE REVIEW

**Comment by sledorze** [comment]
@gcanti before involving more time in it, I wanted to have your feedback on it.
This would also adds `ReadTask`.

**Comment by sledorze** [comment]
@gcanti I get your thinking.
Last input from me before we drop; regardless of failure handling, I think it may have some value even without possible failure.
We may keep it for DX or drop it as it may be considered trivial to achieve; let me know.

Other than that, I may propose to add all the overload of the new 'xxx3C' introduced in this PR; ok to have it inside this PR (or another one?)

I've done it in the previous commit for Applicative3C (necessary for traverse test)

Remaining:
Alt3C
Apply3C
Chain3C
Functor3C
Monad3C

May I add those in this PR?

**Comment by gcanti** [comment]
> We may keep it for DX or drop it as it may be considered trivial to achieve; let me know

I think they are confusing, let's drop them

> May I add those in this PR?

ok

File: src/Alt.ts [hunk 1]
```diff
@@ -32,6 +32,13 @@ export interface Alt2<F extends URIS2> extends Functor2<F> {
   readonly alt: <E, A>(fx: Kind2<F, E, A>, fy: () => Kind2<F, E, A>) => Kind2<F, E, A>
 }
 
+/**
+ * @since 2.0.0
```
Comment (gcanti) [inline src/Alt.ts]: should be `@since 2.2.0` (same for the other 3C interfaces..)

Comment (sledorze) [reply src/Alt.ts]: Ok it will then be 2.2.0 :)

**Comment by sledorze** [comment]
I can't find why the coverage is not 100% on ReaderTaskEither (242) everything is covered to me

**Comment by sledorze** [comment]
@gcanti 
I would eat a raw habanero to have 2.2.0 released by today (and may help) :)

**Comment by mikearnaldi** [comment]
@sledorze @gcanti I'll gladly share the habanero with stephan trasversable on matechs-effect depends on this :)

**Comment by sledorze** [comment]
@gcanti I know it's a big one, if that's an issue maybe a 'next' pre-release?

**Comment by sledorze** [comment]
@gcanti updated for 2.3

File: src/ReaderTask.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+ * @since 2.3.0
+ */
+export interface ReaderTask<R, A> {
+  (r: R): Task<A>
+}
+
+/**
+ * @since 2.3.0
+ */
+export function run<R, A>(ma: ReaderTask<R, A>, r: R): Promise<A> {
```
Comment (gcanti) [inline src/ReaderTask.ts]: we can drop this function

File: src/ReaderTask.ts [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+/**
+ * @since 2.3.0
+ */
+export const fromTask: <R, A>(ma: Task<A>) => ReaderTask<R, A> = T.fromM
+
+/**
+ * @since 2.3.0
+ */
+export const reader: <R, A = never>(ma: Reader<R, A>) => ReaderTask<R, A> = T.fromReader
```
Comment (gcanti) [inline src/ReaderTask.ts]: should be named `fromReader`

File: src/ReaderTask.ts [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+ * @since 2.3.0
+ */
+export function fromIO<R, A>(ma: IO<A>): ReaderTask<R, A> {
+  return fromTask(TA.fromIO(ma))
+}
+
+/**
+ * @since 2.3.0
+ */
+export function of<R, A>(a: A): ReaderTask<R, A> {
```
Comment (gcanti) [inline src/ReaderTask.ts]: this is

```ts
export const of: <R, A>(a: A) => ReaderTask<R, A> = T.of
```

File: src/ReaderTask.ts [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+  return fromTask(TA.of(a))
+}
+
+/**
+ * @since 2.3.0
+ */
+export function fromReader<R, A>(ma: Reader<R, A>): ReaderTask<R, A> {
```
Comment (gcanti) [inline src/ReaderTask.ts]: duplicated

**Comment by gcanti** [comment]
Thanks @sledorze

**Pull request merged**